client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -166,73 +167,159 @@ public Response performRequest(String method, String endpoint, Map<String, Strin
+                        }
+                    }
+                } catch(Exception e) {
+                    listener.onDefinitiveFailure(e);
                 }
-                //also reset the request to make it reusable for the next attempt
-                request.reset();
             }
 
-            CloseableHttpResponse httpResponse;
-            try {
-                httpResponse = client.execute(host, request);
-            } catch(IOException e) {
-                RequestLogger.logFailedRequest(logger, request, host, e);
-                onFailure(host);
-                lastSeenException = addSuppressedException(lastSeenException, e);
-                continue;
-            }
-            Response response = new Response(request.getRequestLine(), host, httpResponse);
-            int statusCode = response.getStatusLine().getStatusCode();
-            if (statusCode < 300 || (request.getMethod().equals(HttpHead.METHOD_NAME) && statusCode == 404) ) {
-                RequestLogger.logResponse(logger, request, host, httpResponse);
-                onResponse(host);
-                return response;
+            @Override
+            public void failed(Exception failure) {
+                try {
+                    RequestLogger.logFailedRequest(logger, request, host, failure);
+                    onFailure(host);
+                    retryIfPossible(failure, hosts, request);
+                } catch(Exception e) {
+                    listener.onDefinitiveFailure(e);
+                }
             }
-            RequestLogger.logResponse(logger, request, host, httpResponse);
-            String responseBody;
-            try {
-                if (response.getEntity() == null) {
-                    responseBody = null;
+
+            private void retryIfPossible(Exception exception, Iterator<HttpHost> hosts, HttpRequestBase request) {
+                if (hosts.hasNext()) {
+                    //in case we are retrying, check whether maxRetryTimeout has been reached
+                    long timeElapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+                    long timeout = maxRetryTimeoutMillis - timeElapsedMillis;
+                    if (timeout <= 0) {
+                        IOException retryTimeoutException = new IOException(
+                                "request retries exceeded max retry timeout [" + maxRetryTimeoutMillis + "]");
+                        listener.onDefinitiveFailure(retryTimeoutException);
+                    } else {
+                        listener.trackFailure(exception);
+                        request.reset();
+                        performRequest(startTime, hosts, request, responseConsumer, listener);
+                    }
                 } else {
-                    responseBody = EntityUtils.toString(response.getEntity());
+                    listener.onDefinitiveFailure(exception);
```

Comment:
I see, so it could be beneficial to have in the response a list of all the retried hosts for instance, rather than only the one that succeded. we can do that later I think.